src/scatter.jl

```diff
@@ -0,0 +1,193 @@
+
+# Arguments
+- `ys`: the destination for `us` to aggregate to. This argument will be mutated.
+- `us`: the source data for aggregating.
+- `xs`: the mapping for aggregation from source (index) to destination (value).
+The index of `xs` is corresponding to the index of `us` and the value of `xs` is
+corresponding to the index of `ys`. The value of `xs` can be `Int` or `Tuple` type.
+
+The dimension of `us` must equal to dimension of `xs`. `ys`, `us` and `xs` must be
+supported array type and be the same type.`Array`, `StaticArray` and `CuArray`
+are currently supported.
+"""
+function scatter_add! end
+
+"""
+    scatter_sub!(ys, us, xs)
+
+Scatter subtraction operation.
+
+    ys[xs[k]...] = ys[xs[k]...] .- us[k...]
+
+# Arguments
+- `ys`: the destination for `us` to aggregate to. This argument will be mutated.
+- `us`: the source data for aggregating.
+- `xs`: the mapping for aggregation from source (index) to destination (value).
+The index of `xs` is corresponding to the index of `us` and the value of `xs` is
+corresponding to the index of `ys`. The value of `xs` can be `Int` or `Tuple` type.
+
+The dimension of `us` must equal to dimension of `xs`. `ys`, `us` and `xs` must be
+supported array type and be the same type.`Array`, `StaticArray` and `CuArray`
+are currently supported.
+"""
+function scatter_sub! end
+
+"""
+    scatter_mul!(ys, us, xs)
+
+Scatter multiplication operation.
+
+    ys[xs[k]...] = ys[xs[k]...] .* us[k...]
+
+# Arguments
+- `ys`: the destination for `us` to aggregate to. This argument will be mutated.
+- `us`: the source data for aggregating.
+- `xs`: the mapping for aggregation from source (index) to destination (value).
+The index of `xs` is corresponding to the index of `us` and the value of `xs` is
+corresponding to the index of `ys`. The value of `xs` can be `Int` or `Tuple` type.
+
+The dimension of `us` must equal to dimension of `xs`. `ys`, `us` and `xs` must be
+supported array type and be the same type.`Array`, `StaticArray` and `CuArray`
+are currently supported.
+"""
+function scatter_mul! end
+
+"""
+    scatter_div!(ys, us, xs)
+
+Scatter division operation.
+
+    ys[xs[k]...] = ys[xs[k]...] ./ us[k...]
+
+# Arguments
+- `ys`: the destination for `us` to aggregate to. This argument will be mutated.
+- `us`: the source data for aggregating.
+- `xs`: the mapping for aggregation from source (index) to destination (value).
+The index of `xs` is corresponding to the index of `us` and the value of `xs` is
+corresponding to the index of `ys`. The value of `xs` can be `Int` or `Tuple` type.
+
+The dimension of `us` must equal to dimension of `xs`. `ys`, `us` and `xs` must be
+supported array type and be the same type.`Array`, `StaticArray` and `CuArray`
+are currently supported.
+"""
+function scatter_div! end
+
+for op = [:add, :sub, :mul, :div]
+    fn = Symbol("scatter_$(op)!")
+    @eval function $fn(ys::Array{T}, us::Array{T}, xs::Array{<:IntOrTuple}) where {T<:Real}
```

Comment:
Can the signatures be relaxed to AbstractArrays?

Reply:
Done.